Tape Recorder

WhiteHusky opened the pull request:
# Description

This adds a voice tape recorder that's able to record nearby speech and then play it back, maintaining the language the person spoke in the playback. Later in this PR will implement cassettes to swap between into the recorder.

---

# TODO

- [x] Sprite changing between Idle, Play, and Record
- [x] Cassettes
- [x] Sounds
- [x] Recording limits
- [ ] Test the following situations, including languages:
  - [ ] Playback alone
  - [ ] Playback into a radio
  - [ ] Playback into a camera (onto TVs)
  - [ ] Playback with multiple recorders
  - [ ] Record alone
  - [ ] Record from a radio
  - [ ] Record from a TV
  - [ ] Record simple mobs
  - [ ] Record with multiple recorders

---
<details><summary><h1>Media</h1></summary>
<p>

[Skip Silence Feature](https://github.com/user-attachments/assets/997ce56a-23d2-46a7-8d43-831c83e6f100)

</p>
</details>

---

# Changelog

<!--
You can add an author after the `:cl:` to change the name that appears in the changelog (ex: `:cl: Death`)
Leaving it blank will default to your GitHub display name
This includes all available types for the changelog
-->

:cl:Carlen White
- add: Voice Tape Recorder

WhiteHusky committed:
Prototype Tape Recorder

LucasTheDrgn commented:
One thing I will preemptively mention is that TVs DON'T respect the language system as-is. Anything said into a TV will come out in, essentially, universal. This might cause some weirdness with recording from TVs, and might cause some confusion when playing back through TVs. Other than that, I wish you luck and godspeed!

WhiteHusky committed:
Readability cleanup

WhiteHusky committed:
ScheduleNextRecorder on every playing recorder

Avoid a situation where several recorders are playing and the one scheduled to say something first has stopped and thus all other recorders are unable to continue.

WhiteHusky committed:
Change order of alternate verbs; volume then erase

Destructive actions *should* be lowest priority.

WhiteHusky committed:
Change the field name NormalVolume to HighVolume

WhiteHusky committed:
Hide the Erase verb if the player is doing something

WhiteHusky committed:
Audio, Cassettes, and Silence Skip implemented

Also fixed/added some attributions

WhiteHusky committed:
Correct attribution

WhiteHusky committed:
Adjust volumes

WhiteHusky committed:
Implement recording limits

WhiteHusky commented:
PR is basically almost done and just need to poke and prod for bugs while getting some opinions over the code.
The only thing I need to do is make cassette swapping a bit more fluid since the ejected cassette will just spill onto the floor, but that's a matter of looking at how flashlights handle battery cell swaps and such.

WhiteHusky commented:
@LucasTheDrgn I might end up fixing the TV in a separate PR since it's uniquely broken and brings it out of scope of this PR. The main issue is that anything that needs to handle language, especially if relaying like the TV does, need to be _aware_ of the language system and do things in a specific way to maintain the language heard.

This PR _really_ abuses the language system as-is which allows the tape recorder to properly imitate people and languages by both having a `LanguageSpeakerComponent` and `VoiceOverrideComponent` attached to the tape recorder so things work correctly.

WhiteHusky committed:
Fix out of date comment

WhiteHusky committed:
Formatting

WhiteHusky committed:
Optimize Update method

WhiteHusky committed:
Rename variables and methods

WhiteHusky committed:
Simplify RecorderState checks

WhiteHusky committed:
Specify item sizes

WhiteHusky committed:
Have `ScheduleNextRecorder` return a struct
Also rename `component` to `recorder` as needed

WhiteHusky committed:
Use ItemSlot for the cassette instead

WhiteHusky committed:
Allow stashing tiny items inside the player

WhiteHusky committed:
Fix attempting to record with alt-click causing a Record => Idle => Record mess

WhiteHusky commented:
Pretty much at Release Candidate status now; it *largely* works how I want.

However there's a bug I need to solve: Alt-Clicking the player with a inserted cassette will fake-out a eject, but it will Record/Idle normally.
Checking the code for other things with `ItemSlots`, like the CryoPod, this _seems_ to be fixed by using a Client/Shared system which I guess allows client prediction to handle it gracefully. I'm not looking forward doing that, but I must it seems.

WhiteHusky commented:
[I guess my abuse of the slot fails the test.](https://github.com/Fansana/floofstation1/actions/runs/14320634458/job/40143167357?pr=665#step:10:131)

WhiteHusky commented:
Okay so I'm stuck on this bizzaro bug when trying to solve that implementation issue by manually handling the eject myself.
![image](https://github.com/user-attachments/assets/476cfabc-921e-4a1d-a91e-2496cc3aee78)
That second entry _shouldn't even be there_ and it's being created by something else. Using that bottom entry does a fake-eject before snapping the cassette out of the user's hand while the upper one functions as expected.

ghost commented:
Seems like there a re a bunch of failed tests, if you need help fixing those you can ping me on discord :)

WhiteHusky commented:
> Seems like there a re a bunch of failed tests, if you need help fixing those you can ping me on discord :)

Oh the issue is a bit more deeper. At the moment I plan to do break the monolithic `Server.VoiceTapeRecorderSystem` into also `Shared.SharedVoiceTapeRecorderSystem` and `Client.VoiceTapeRecorderSystem` to hopefully fix the insertion issue while making the recorder feel snappier on the client.

Eagle0600 commented:
Are you aware that DeltaV has a fully working tape recorder, with transcript printing, a nice UI, and etcetera? We could port that.

Although, one thing I will say that I want from tape recorders that DeltaV doesn't have is an easy way to erase a tape. Currently, you can only erase a tape by recording over it, which requires two minutes of silence.
